op-node/rollup/engine/engine_reset.go

```diff
@@ -7,72 +7,54 @@ import (
 	"github.com/ethereum/go-ethereum/log"
 
 	"github.com/ethereum-optimism/optimism/op-node/rollup"
-	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
 	"github.com/ethereum-optimism/optimism/op-node/rollup/sync"
-	"github.com/ethereum-optimism/optimism/op-service/eth"
 )
 
-type ResetL2 interface {
-	sync.L2Chain
-	derive.SystemConfigL2Fetcher
-}
+// ResetEngineRequestEvent requests the EngineResetDeriver to walk
+// the L2 chain backwards until it finds a plausible unsafe head,
+// and find an L2 safe block that is guaranteed to still be from the L1 chain.
+type ResetEngineRequestEvent struct{}
 
-type ResetEngineControl interface {
-	SetUnsafeHead(eth.L2BlockRef)
-	SetSafeHead(eth.L2BlockRef)
-	SetFinalizedHead(eth.L2BlockRef)
+func (ev ResetEngineRequestEvent) String() string {
+	return "reset-engine-request"
+}
 
-	SetBackupUnsafeL2Head(block eth.L2BlockRef, triggerReorg bool)
-	SetPendingSafeL2Head(eth.L2BlockRef)
+type EngineResetDeriver struct {
+	ctx     context.Context
+	log     log.Logger
+	cfg     *rollup.Config
+	l1      sync.L1Chain
+	l2      sync.L2Chain
+	syncCfg *sync.Config
 
-	ResetBuildingState()
+	emitter rollup.EventEmitter
 }
 
-// ResetEngine walks the L2 chain backwards until it finds a plausible unsafe head,
-// and an L2 safe block that is guaranteed to still be from the L1 chain.
-func ResetEngine(ctx context.Context, log log.Logger, cfg *rollup.Config, ec ResetEngineControl, l1 sync.L1Chain, l2 ResetL2, syncCfg *sync.Config, safeHeadNotifs rollup.SafeHeadListener) error {
-	result, err := sync.FindL2Heads(ctx, cfg, l1, l2, log, syncCfg)
-	if err != nil {
-		return derive.NewTemporaryError(fmt.Errorf("failed to find the L2 Heads to start from: %w", err))
-	}
-	finalized, safe, unsafe := result.Finalized, result.Safe, result.Unsafe
-	l1Origin, err := l1.L1BlockRefByHash(ctx, safe.L1Origin.Hash)
-	if err != nil {
-		return derive.NewTemporaryError(fmt.Errorf("failed to fetch the new L1 progress: origin: %v; err: %w", safe.L1Origin, err))
-	}
-	if safe.Time < l1Origin.Time {
-		return derive.NewResetError(fmt.Errorf("cannot reset block derivation to start at L2 block %s with time %d older than its L1 origin %s with time %d, time invariant is broken",
-			safe, safe.Time, l1Origin, l1Origin.Time))
```

Comment:
Flagging that we seem to have lost this check along the way. I'm not overly concerned about that given it should never happen and the node is pretty unlikely to recover and will just wind up in a reset loop because of this. We don't typically try to reverify blocks we've already imported and this time invariant would have been checked on the block import anyway.

Reply:
This code was duplicated in a previous iteration: the pipeline now deals with turning the L2 safe head into the L1 origin that is being started from. All it was doing here was fetching and logging, so I dropped it for simplicity.